server/swap/swap_test.go

```diff
@@ -747,9 +748,25 @@ func (rig *testRig) checkMatchNotification(msg *msgjson.Message, oid order.Order
 	return nil
 }
 
+// helper to check the swap status for specified user.
+func (rig *testRig) checkSwapStatusChange(wantStatus order.MatchStatus, user *tUser) error {
+	// Swap counterparty ultimately gets an audit request. Synchronize with that.
+	if err := rig.waitChans("counterparty received our swap -> counterparty audit request", nil, rig.auth.swapReceived, rig.auth.auditReq); err != nil {
+		return err
+	}
+	tracker := rig.getTracker()
+	tracker.mtx.RLock()
+	status := tracker.Status
+	tracker.mtx.RUnlock()
```

Comment:
Seems like there are a bunch of places in these tests where mutex should be used, like this one,
if reviewers agree, I'd add the required mutex calls I've spotted in this test-file

Reply:
and I've actually encountered a race couple of times while running these tests locally:

```
==================
WARNING: DATA RACE
Write at 0x00c0004be348 by goroutine 38:
  decred.org/dcrdex/server/swap.(*Swapper).processInit()
      /Users/iurii/dcrdex/server/swap/swap.go:1618 +0x1c07
  decred.org/dcrdex/server/swap.(*Swapper).handleInit.func1()
      /Users/iurii/dcrdex/server/swap/swap.go:1957 +0x6b
  decred.org/dcrdex/dex/wait.(*TaperingTickerQueue).Run.func1()
      /Users/iurii/dcrdex/dex/wait/queue.go:197 +0xfa
  decred.org/dcrdex/dex/wait.(*TaperingTickerQueue).Run·dwrap·5()
      /Users/iurii/dcrdex/dex/wait/queue.go:248 +0x47

Previous read at 0x00c0004be348 by goroutine 91:
  decred.org/dcrdex/server/swap.(*testRig).sendSwap_maker()
      /Users/iurii/dcrdex/server/swap/swap_test.go:777 +0x384
  decred.org/dcrdex/server/swap.TestRetrySwapInit()
      /Users/iurii/dcrdex/server/swap/swap_test.go:2041 +0x2de
  testing.tRunner()
      /Users/iurii/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /Users/iurii/go/src/testing/testing.go:1306 +0x47
```

Reply:
Interesting ... haven't thought about increasing test "coverage" this way, makes sense; so I see 2 options then:
- leave other tests as is (maybe I can go over those to see whether I can spot any potential concurrency issues and raise if any), preserving this "coverage", letting race detector do its job for us (and address these, if any, on a need-to basis)
- or (lazy approach I've suggested above) use mutex blindly, basically giving up on this additional "coverage" at the benefit of not having to worry about writing / maintaining precise tests with respect to concurrent access

I can see the benefits of either of these (personally I'm used to option 2) and I'm onboard with whichever you prefer

Reply:
If there's a place where you can reason out a race even if it's not detected, use the mutex.

We're sort of used to the lazy approach in much of the product code, and I can say that it bites us.   It's very important to have a good understanding of access patterns because over using mutexes can be bad for not only performance but it can get you into deadlocks and undesirable mutex contention.

Reply:
On this note, while I think it's good to be acutely aware of what goroutines are running and why it is or isn't safe to access some field, I'm no longer fond of the general approach in our tests of being so deeply involved with internals and implementation details rather than testing the API.  The fact that we have to do this is partly a consequence of the need to stub out connections and databases, but it's also a hint that we've not factored out enough logic into smaller functions so we can test them in isolation, among other things like wanting to reuse structures instead of making fresh ones, doing more repeated teardown/bringup, etc.

It's frustrating to change some small detail in the implementation that does not affect the overall function of the system but ends up completely wrecking a test that expects some irrelevant order of operations or some message that we hack a channel into via a stub, etc.

This is not a comment that you should change anything, just saying that while I always prefer a developer to think hard about access patterns and if a mutex is even needed, I generally do not love our overall testing approach.  This is an evolving view of mine.

Reply:
I've looked a bit more at how `tracker.Status` is accessed, (in `TestRetrySwapInit`) it gets synchronized on `rig.auth.swapReceived` (or for redeems it would be `rig.auth.redeemReceived`) with the concurrent Go-routine that is running `processInit`, thus locking `tracker.mtx` isn't really necessary (so, I removed locking).
In other tests there seems to be no concurrent access at all.